subsys/usb/class/bt_h4.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Wireless / Bluetooth USB class
+ *
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <sys/byteorder.h>
+
+#include <usb/usb_device.h>
+#include <usb/usb_common.h>
+#include <usb_descriptor.h>
+
+#include <net/buf.h>
+
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+#include <bluetooth/l2cap.h>
+
+#define LOG_LEVEL CONFIG_USB_DEVICE_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_bt_h4);
+
+static K_FIFO_DEFINE(rx_queue);
+static K_FIFO_DEFINE(tx_queue);
+
+/* ACL data TX buffers */
+#if defined(CONFIG_BT_CTLR_TX_BUFFERS)
+#define BUF_COUNT (CONFIG_BT_HCI_CMD_COUNT + CONFIG_BT_CTLR_TX_BUFFERS)
+#else
+#define BUF_COUNT (CONFIG_BT_HCI_CMD_COUNT + 4)
+#endif
+
+#if defined(CONFIG_BT_CTLR_TX_BUFFER_SIZE)
+#define BT_L2CAP_MTU (CONFIG_BT_CTLR_TX_BUFFER_SIZE - BT_L2CAP_HDR_SIZE)
+#else
+#define BT_L2CAP_MTU 64
+#endif
+
+/* Data size needed for CMD and ACL buffers */
+#define BUF_SIZE MAX(BT_BUF_RX_SIZE, BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU))
+NET_BUF_POOL_DEFINE(rx_pool, BUF_COUNT, BUF_SIZE, sizeof(u8_t), NULL);
+
+#define BT_H4_OUT_EP_ADDR               0x01
+#define BT_H4_IN_EP_ADDR                0x81
+
+#define BT_H4_OUT_EP_IDX                0
+#define BT_H4_IN_EP_IDX                 1
+
+#define BT_H4_BULK_EP_MPS               MIN(BUF_SIZE, USB_MAX_FS_BULK_MPS)
+
+/* HCI RX/TX threads */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, 512);
```

Comment:
Should these come from Kconfig?

Reply:
Don't think one would need to change the stack size, that is why there are dedicated threads instead of using something like the syswq.

Reply:
The RX thread will call bt_send which will call `bt_dev.drv->send`. We have the possebility to register different hci_drivers here, and those drivers might be (as in my case) out of tree drivers.

Reply:
How it is working now then? It doesn't seem to be configurable in bluetooth.c either, why would I  have to assume this would be different to bt_h4, if you have a requirement not attended by the stack please send a patch since it is kind counter productive for me to guess how your out of the tree driver would work I may not even guess what default and range would work best which is another reason to have this pushed as soon as possible so I don't have to resolve every driver problem in one, never ending, PR. For the most part this is just an example for how H4 mode would work as stand alone mode, don't think anyone would be thinking on productize it without it being part of the Bluetooth spec.

Reply:
Right. Then it doesn't work in that case for bluetooth.c either. I'll make a note of it and and fix it for our out of tree driver afterwards if needed. So just leave it as is.